libs/editor/egui_editor/src/bounds.rs

```diff
@@ -514,12 +553,188 @@ impl DocCharOffset {
     }
 }
 
+pub trait RangesExt {
+    /// Efficiently finds the possibly empty (inclusive, exclusive) range of ranges that contain `offset`
+    fn find_containing(
+        &self, offset: DocCharOffset, start_inclusive: bool, end_inclusive: bool,
+    ) -> (usize, usize);
+
+    /// Efficiently finds the possibly empty (inclusive, exclusive) range of ranges that intersect `range`
+    fn find_intersecting(
+        &self, range: (DocCharOffset, DocCharOffset), allow_empty: bool,
+    ) -> (usize, usize);
+}
```

Comment:
factoring/performance: some new generic helper functions for working with bounds; they use binary search instead of linear